Gauss laguerre

ckormanyos opened the pull request:
Remove unused variables.
Check with GCC's -Wall -Wextra -pedantic and a few more warnings.
Remove all MSVC level 3 warnings.
Other compiler warnings not checked.
Eliminate unused logic paths.
Speedup and calibrate for 50...1200 decimal digits.
Use a few more Boost macros, but not yet fully constexpr.
Repair the correct author and update the copyright information.

ckormanyos commented:
This PR is intended to implement the first round or repairs and improvements of the Gauss Laguerre quadrature example from 2012-2015.

Some of these points have been mentioned in #193

ckormanyos commented:
Actually, I would like to extend this example.
Will close this PR without merging it.

ckormanyos commented:
The Gauss-Laguerre quadrature example has been reworked.
The approximation now handles arguments on the real axis airy_ai(x) with x.gt.1.
The example now checks a range of arguments with control values from `cyl_bessel_k` of 1/3.
The idea of this example is now much more clear, as mentioned in code comments.

- This example is intended to handle airy_ai(x) calculations for medium to large arguments
- Use quadrature for x >=1
- Assume that and a hypergeometric function or a Taylor series could be used for x < 1.

I will probably merge my own PR in a day or so, assuming the tests run.

NAThompson commented:
Very cool. Just a few comments: I added a progress bar into:

https://github.com/boostorg/math/blob/develop/example/naive_monte_carlo_example.cpp

It seems like this example might also benefit from the `display_progress` function, since it takes a while to run, with no output for some time, and then lots of output which could potentially be better expressed as a progress bar.

I ran this example through the Sanitizer Suite over both clang++-9 and g++-7; it came out clean. So looks good to me!

NAThompson commented on example/gauss_laguerre_quadrature.cpp:
```diff
@@ -26,258 +31,222 @@
-                                  * alpha_pow_minus_two_thirds + 1.0F);
+        first_laguerre_root = (j_alpha_m1_sqr * (   -T(0.6666666666667F)
+                                                 + ((T(0.6666666666667F) * alpha) * alpha)
+                                                 +  (T(0.3333333333333F) * j_alpha_m1_sqr) + vf2)) / (vf2 * vf);
```
Doesn't this lose precision? I might do T(2)/T(3), or boost::math::constants::two_thirds.

pabristow commented:
I pondered on this too, but concluded that it is only a guess, so double is more than good enough?

Looks good (if it took a little while for a superannuated FORTRAN programmer to et his head around some of the lambda-ing).  While this is indubitably cool, is there a performance reason for doing it this way?

From: Nick <notifications@github.com>
Sent: 22 February 2020 17:45
To: boostorg/multiprecision <multiprecision@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [boostorg/multiprecision] Gauss laguerre (#196)


@NAThompson commented on this pull request.

________________________________

In example/gauss_laguerre_quadrature.cpp<https://github.com/boostorg/multiprecision/pull/196#discussion_r382929140>:

> -      }

-      else

-      {

-        // For larger alpha, use the first line of Eqs. 10.21.40 in the NIST Handbook.

-        const T alpha_pow_third(boost::math::cbrt(alpha));

-        const T alpha_pow_minus_two_thirds(T(1) / (alpha_pow_third * alpha_pow_third));

-

-        j_alpha_m1 = alpha * (((((                             + 0.043F

-                                  * alpha_pow_minus_two_thirds - 0.0908F)

-                                  * alpha_pow_minus_two_thirds - 0.00397F)

-                                  * alpha_pow_minus_two_thirds + 1.033150F)

-                                  * alpha_pow_minus_two_thirds + 1.8557571F)

-                                  * alpha_pow_minus_two_thirds + 1.0F);

+        first_laguerre_root = (j_alpha_m1_sqr * (   -T(0.6666666666667F)

+                                                 + ((T(0.6666666666667F) * alpha) * alpha)

+                                                 +  (T(0.3333333333333F) * j_alpha_m1_sqr) + vf2)) / (vf2 * vf);

Doesn't this lose precision? I might do T(2)/T(3), or boost::math::constants::two_thirds.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/boostorg/multiprecision/pull/196?email_source=notifications&email_token=AAIG4ALUEXBVWDD4XKKKTDTREFQDBA5CNFSM4KWFWCJ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCWRZEZI#pullrequestreview-363041381>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAIG4AKHF6M5DRHDHFTJHMTREFQDBANCNFSM4KWFWCJQ>.

NAThompson commented:
Whoops, didn't notice that it's a guess; my bad.

ckormanyos commented:
> It seems like this example might also benefit from the display_progress function, since it takes a while to run, with no output for some time, and then lots of output which could potentially be better expressed as a progress bar.

Kewl. I will check it out...

I think this example is progressing to be about where I would like it to be.

Based on some new knowledge that I acquired in this area and extension of the example, I think it will make sense that I add some further sentences and an equation or two to the docs, as time permits.

Thanks for the help, and kind regards, Chris

ckormanyos commented:
Is there a simple reason anyone knows why the builds are failing on CI?

I can't see how my changes caused these failures. But I am still investigating...

ckormanyos commented:
>> It seems like this example might also benefit from the display_progress function, since it takes a while to run, with no output for some time, and then lots of output which could potentially be better expressed as a progress bar.

> Kewl. I will check it out...

Progress percent indications have been added.

NAThompson commented:
@ckormanyos : Just tried it out; looks great!

ckormanyos commented:
> Testing especially looking much better.

Yup. Thank you. It was a good idea from your side to clean up this example. I learned a lot and I think the example is much better now.

> I noted this remaining comment...

Which is now expressed differently, thereby removing the open endedness of the comment.

More detailed comments describing the example are also now near the head of the file.

> As far as I can see all the results are 'exact', so perhaps the guard digits are a little over-enthusiastic?

I have removed the guard digits. It was necessary to slightly increase the tolerance of close fraction comparison with the control values. But I do, in fact, prefer this now without the guard digits. It's a bit cleaner to describe and comprehend.

The tests are now cycling. I had a few problems with using some C++11 stuff on the appveyor runs. But I switched over to the Boost macros such as those for older compilers lacking some C++11 features.  Barring any unexpected failures, I will handle this PR myself after the tests cycle.

Thanks and kind regards, Chris

ckormanyos commented:
The Gauss-Laguerre quadrature example has been refactored and improved.
